File: packages/cad/bakeoff/invariants.mjs
// The JS mirror of engine/src/invariants.rs, for meshes that come out of the
// foreign kernels. Same numbers, same definitions; cube.selftest asserts it.
export function weld(pos, tris, tol) {
  const q = (x) => Math.round(x / tol);
  const map = new Map();
  const out = [];
  const remap = new Uint32Array(pos.length);
  pos.forEach((p, i) => {
    const key = `${q(p[0])},${q(p[1])},${q(p[2])}`;
    let idx = map.get(key);
    if (idx === undefined) { idx = out.length; out.push(p); map.set(key, idx); }
    remap[i] = idx;
  });
  const t2 = [];
  for (const t of tris) { const a = remap[t[0]], b = remap[t[1]], c = remap[t[2]]; if (a !== b && b !== c && a !== c) t2.push([a, b, c]); }
  return { pos: out, tris: t2 };
}

export function compute(mesh) {
  const { pos, tris } = mesh;
  let volume = 0, area = 0; const cx = [0, 0, 0];
  const bbox = [[Infinity, Infinity, Infinity], [-Infinity, -Infinity, -Infinity]];
  for (const p of pos) for (let i = 0; i < 3; i++) { bbox[0][i] = Math.min(bbox[0][i], p[i]); bbox[1][i] = Math.max(bbox[1][i], p[i]); }
  const dir = new Map(), uses = new Map();
  for (const t of tris) {
    const a = pos[t[0]], b = pos[t[1]], c = pos[t[2]];
    const cr = [b[1] * c[2] - b[2] * c[1], b[2] * c[0] - b[0] * c[2], b[0] * c[1] - b[1] * c[0]];
    const v = (a[0] * cr[0] + a[1] * cr[1] + a[2] * cr[2]) / 6;
    volume += v; for (let i = 0; i < 3; i++) cx[i] += v * (a[i] + b[i] + c[i]) / 4;
    const ab = [b[0] - a[0], b[1] - a[1], b[2] - a[2]], ac = [c[0] - a[0], c[1] - a[1], c[2] - a[2]];
    const n = [ab[1] * ac[2] - ab[2] * ac[1], ab[2] * ac[0] - ab[0] * ac[2], ab[0] * ac[1] - ab[1] * ac[0]];
    area += Math.hypot(n[0], n[1], n[2]) / 2;
    for (let k = 0; k < 3; k++) { const p = t[k], q = t[(k + 1) % 3]; const key = p < q ? p * 4294967296 + q : q * 4294967296 + p; dir.set(key, (dir.get(key) || 0) + (p < q ? 1 : -1)); uses.set(key, (uses.get(key) || 0) + 1); }
  }
  let open_edges = 0, flipped_edges = 0;
  for (const [k, u] of uses) { if (u !== 2) open_edges++; else if (dir.get(k) !== 0) flipped_edges++; }
  const used = new Uint8Array(pos.length); for (const t of tris) for (const i of t) used[i] = 1;
  let verts = 0; for (const u of used) verts += u;
  const euler = verts - uses.size + tris.length;
  if (Math.abs(volume) > 1e-12) for (let i = 0; i < 3; i++) cx[i] /= volume;
  return { volume, area, bbox, centroid: cx, euler, watertight: tris.length > 0 && open_edges === 0 && flipped_edges === 0, tris: tris.length, verts, open_edges, flipped_edges };
}

export function readStl(buf) {
  const n = buf.readUInt32LE(80); const pos = []; const tris = [];
  for (let i = 0; i < n; i++) { const o = 84 + i * 50; const t = []; for (let k = 1; k < 4; k++) { pos.push([buf.readFloatLE(o + k * 12), buf.readFloatLE(o + k * 12 + 4), buf.readFloatLE(o + k * 12 + 8)]); t.push(pos.length - 1); } tris.push(t); }
  return { pos, tris };
}

export function writeStl(mesh) {
  const { pos, tris } = mesh; const buf = Buffer.alloc(84 + tris.length * 50); buf.writeUInt32LE(tris.length, 80);
  tris.forEach((t, i) => { const o = 84 + i * 50; const a = pos[t[0]], b = pos[t[1]], c = pos[t[2]]; const ab = [b[0] - a[0], b[1] - a[1], b[2] - a[2]], ac = [c[0] - a[0], c[1] - a[1], c[2] - a[2]]; const n = [ab[1] * ac[2] - ab[2] * ac[1], ab[2] * ac[0] - ab[0] * ac[2], ab[0] * ac[1] - ab[1] * ac[0]]; const l = Math.hypot(...n) || 1; [n[0] / l, n[1] / l, n[2] / l, ...a, ...b, ...c].forEach((x, k) => buf.writeFloatLE(x, o + k * 4)); });
  return buf;
}

if (process.argv[1] && process.argv[1].endsWith('invariants.mjs')) {
  const p = (x, y, z) => [x, y, z];
  const pos = [p(0, 0, 0), p(1, 0, 0), p(1, 1, 0), p(0, 1, 0), p(0, 0, 1), p(1, 0, 1), p(1, 1, 1), p(0, 1, 1)];
  const quads = [[0, 3, 2, 1], [4, 5, 6, 7], [0, 1, 5, 4], [1, 2, 6, 5], [2, 3, 7, 6], [3, 0, 4, 7]];
  const tris = []; for (const q of quads) { tris.push([q[0], q[1], q[2]], [q[0], q[2], q[3]]); }
  const inv = compute({ pos, tris });
  const ok = Math.abs(inv.volume - 1) < 1e-12 && Math.abs(inv.area - 6) < 1e-12 && inv.euler === 2 && inv.watertight;
  console.log(ok ? '✓ invariants.mjs cube' : '✗ invariants.mjs cube', inv);
  process.exit(ok ? 0 : 1);
}

File: packages/cad/engine/src/invariants.rs
//! What a test asserts on. Mesh bits change with tessellation and kernel
//! version; a bracket's volume does not. Everything here is computed from a
//! plain triangle soup so it is the same number for every kernel.

use crate::kernel::TriMesh;
use serde::Serialize;
use std::collections::HashMap;

#[derive(Debug, Clone, Serialize)]
pub struct Invariants {
    pub volume: f64,
    pub area: f64,
    pub bbox: [[f64; 3]; 2],
    pub centroid: [f64; 3],
    pub euler: i64,
    pub watertight: bool,
    pub tris: usize,
    pub verts: usize,
    /// edges not used by exactly two triangles
    pub open_edges: usize,
    /// edges whose two triangles disagree on orientation
    pub flipped_edges: usize,
}

fn cross(a: [f64; 3], b: [f64; 3]) -> [f64; 3] {
    [a[1] * b[2] - a[2] * b[1], a[2] * b[0] - a[0] * b[2], a[0] * b[1] - a[1] * b[0]]
}
fn dot(a: [f64; 3], b: [f64; 3]) -> f64 { a[0] * b[0] + a[1] * b[1] + a[2] * b[2] }
fn sub(a: [f64; 3], b: [f64; 3]) -> [f64; 3] { [a[0] - b[0], a[1] - b[1], a[2] - b[2]] }

pub fn compute(m: &TriMesh) -> Invariants {
    let mut volume = 0.0;
    let mut area = 0.0;
    let mut cx = [0.0; 3];
    let mut bbox = [[f64::INFINITY; 3], [f64::NEG_INFINITY; 3]];
    for p in &m.pos {
        for i in 0..3 {
            bbox[0][i] = bbox[0][i].min(p[i]);
            bbox[1][i] = bbox[1][i].max(p[i]);
        }
    }
    // signed tetra volumes from the origin; centroid of volume
    let mut edges: HashMap<(u32, u32), i32> = HashMap::new();
    for t in &m.tris {
        let (a, b, c) = (m.pos[t[0] as usize], m.pos[t[1] as usize], m.pos[t[2] as usize]);
        let v = dot(a, cross(b, c)) / 6.0;
        volume += v;
        for i in 0..3 {
            cx[i] += v * (a[i] + b[i] + c[i]) / 4.0;
        }
        let n = cross(sub(b, a), sub(c, a));
        area += dot(n, n).sqrt() / 2.0;
        for k in 0..3 {
            let (p, q) = (t[k], t[(k + 1) % 3]);
            let key = if p < q { (p, q) } else { (q, p) };
            let e = edges.entry(key).or_insert(0);
            // +1 for one direction, -1 for the other; a watertight, consistently
            // oriented mesh has every edge summing to 0 with exactly 2 uses.
            *e += if p < q { 1 } else { -1 };
        }
    }
    let mut uses: HashMap<(u32, u32), u32> = HashMap::new();
    for t in &m.tris {
        for k in 0..3 {
            let (p, q) = (t[k], t[(k + 1) % 3]);
            let key = if p < q { (p, q) } else { (q, p) };
            *uses.entry(key).or_insert(0) += 1;
        }
    }
    let open_edges = uses.values().filter(|&&u| u != 2).count();
    let flipped_edges = edges.iter().filter(|(k, &s)| uses[k] == 2 && s != 0).count();
    let watertight = !m.tris.is_empty() && open_edges == 0 && flipped_edges == 0;
    let verts = {
        let mut used = vec![false; m.pos.len()];
        for t in &m.tris {
            for &i in t {
                used[i as usize] = true;
            }
        }
        used.iter().filter(|&&u| u).count()
    };
    let euler = verts as i64 - edges.len() as i64 + m.tris.len() as i64;
    if volume.abs() > 1e-12 {
        for c in &mut cx {
            *c /= volume;
        }
    }
    Invariants { volume, area, bbox, centroid: cx, euler, watertight, tris: m.tris.len(), verts, open_edges, flipped_edges }
}

/// Weld vertices closer than `tol` so invariants (Euler, watertight) are
/// about the geometry rather than the tessellator's bookkeeping.
pub fn weld(m: &TriMesh, tol: f64) -> TriMesh { weld_with(m, tol, &[]).0 }

/// `weld`, also carrying a per-triangle payload (face ids) through the
/// degenerate-triangle drop so the two stay aligned.
pub fn weld_with(m: &TriMesh, tol: f64, per_tri: &[u32]) -> (TriMesh, Vec<u32>) {
    let q = |x: f64| -> i64 { (x / tol).round() as i64 };
    let mut map: HashMap<(i64, i64, i64), u32> = HashMap::new();
    let mut pos = Vec::new();
    let mut remap = vec![0u32; m.pos.len()];
    for (i, p) in m.pos.iter().enumerate() {
        let key = (q(p[0]), q(p[1]), q(p[2]));
        let idx = *map.entry(key).or_insert_with(|| {
            pos.push(*p);
            (pos.len() - 1) as u32
        });
        remap[i] = idx;
    }
    let mut tris = Vec::with_capacity(m.tris.len());
    let mut ids = Vec::new();
    for (i, t) in m.tris.iter().enumerate() {
        let t = [remap[t[0] as usize], remap[t[1] as usize], remap[t[2] as usize]];
        if t[0] != t[1] && t[1] != t[2] && t[0] != t[2] {
            tris.push(t);
            if let Some(&id) = per_tri.get(i) {
                ids.push(id);
            }
        }
    }
    (TriMesh { pos, tris }, ids)
}

/// Binary STL.
pub fn stl(m: &TriMesh) -> Vec<u8> {
    let mut out = Vec::with_capacity(84 + m.tris.len() * 50);
    out.extend_from_slice(&[0u8; 80]);
    out.extend_from_slice(&(m.tris.len() as u32).to_le_bytes());
    for t in &m.tris {
        let (a, b, c) = (m.pos[t[0] as usize], m.pos[t[1] as usize], m.pos[t[2] as usize]);
        let n = cross(sub(b, a), sub(c, a));
        let l = dot(n, n).sqrt().max(1e-300);
        for v in [[n[0] / l, n[1] / l, n[2] / l], a, b, c] {
            for x in v {
                out.extend_from_slice(&(x as f32).to_le_bytes());
            }
        }
        out.extend_from_slice(&[0u8; 2]);
    }
    out
}

#[cfg(test)]
mod tests {
    use super::*;
    #[test]
    fn unit_cube() {
        let p = |x, y, z| [x, y, z];
        let pos = vec![p(0.0, 0.0, 0.0), p(1.0, 0.0, 0.0), p(1.0, 1.0, 0.0), p(0.0, 1.0, 0.0), p(0.0, 0.0, 1.0), p(1.0, 0.0, 1.0), p(1.0, 1.0, 1.0), p(0.0, 1.0, 1.0)];
        let quads = [[0, 3, 2, 1], [4, 5, 6, 7], [0, 1, 5, 4], [1, 2, 6, 5], [2, 3, 7, 6], [3, 0, 4, 7]];
        let mut tris = Vec::new();
        for q in quads {
            tris.push([q[0], q[1], q[2]]);
            tris.push([q[0], q[2], q[3]]);
        }
        let inv = compute(&TriMesh { pos, tris });
        assert!((inv.volume - 1.0).abs() < 1e-12);
        assert!((inv.area - 6.0).abs() < 1e-12);
        assert_eq!(inv.euler, 2);
        assert!(inv.watertight);
        assert!((inv.centroid[0] - 0.5).abs() < 1e-12);
    }
}
